fedimint-client-legacy/src/mint/mod.rs

```diff
@@ -383,13 +384,87 @@ impl MintClient {
     pub async fn select_notes(&self, amount: Amount) -> Result<TieredMulti<SpendableNote>> {
-        let notes = self.notes().await;
-        let selected_notes = notes.select_notes(amount).ok_or_else(|| {
-            MintClientError::InsufficientBalance(amount, TieredMulti::total_amount(&notes))
-        })?;
+        let mut t = self.start_dbtx().await;
+        let note_stream = t
+            .find_by_prefix_sorted_descending(&NoteKeyPrefix)
+            .await
+            .map(|(key, note)| (key.amount, note));
+        Self::select_notes_from_stream(note_stream, amount).await
+    }
 
-        Ok(selected_notes)
+    async fn select_notes_from_stream<Note>(
+        stream: impl futures::Stream<Item = (Amount, Note)>,
+        requested_amount: Amount,
+    ) -> Result<TieredMulti<Note>> {
+        if requested_amount == Amount::ZERO {
+            return Ok(TieredMulti::default());
+        }
+        let mut stream = Box::pin(stream);
+        let mut selected = vec![];
+        let mut total_amount = Amount::ZERO;
+        let mut last_big_note_checkpoint: Option<(Amount, Note, usize)> = None;
+        let mut pending_amount = requested_amount;
+        let mut previous_amount: Option<Amount> = None; // used to assert descending order
+        while let Some((note_amount, note)) = stream.next().await {
+            assert!(
+                previous_amount.map_or(true, |previous| previous >= note_amount),
+                "notes are not sorted in descending order"
+            );
+            previous_amount = Some(note_amount);
+            total_amount += note_amount;
+            match note_amount.cmp(&pending_amount) {
+                Ordering::Less => {
+                    // keep adding notes until we have enough
+                    pending_amount -= note_amount;
+                    selected.push((note_amount, note))
+                }
+                Ordering::Equal => {
+                    // exactly enough notes, return
+                    selected.push((note_amount, note));
+                    return Ok(selected.into_iter().collect());
+                }
+                Ordering::Greater => {
+                    // probably we don't need this big note, but we'll keep it in case the following
+                    // small notes don't add up to the requested amount
+                    last_big_note_checkpoint = Some((note_amount, note, selected.len()));
+                }
+            }
+        }
+        assert!(pending_amount > Amount::ZERO);
+        if requested_amount > total_amount {
+            return Err(MintClientError::InsufficientBalance(
+                requested_amount,
+                total_amount,
+            ));
+        }
+        let (big_note_amount, big_note, checkpoint) =
+            last_big_note_checkpoint.expect("to have at least one big note");
+        // we can't make exact change, remove all small notes after the big note
+        selected.truncate(checkpoint);
```

Comment:
I'm confused what's going here. Either there's something wrong here, or this function might need a bit more comments to explain things for a bit slower (yet still valuable!) developers like myself. :D

Reply:
Take a look at the new comments